hal/src/nRF52840/timer_hal.cpp

```diff
@@ -15,14 +15,304 @@
+            // the counter.
+            sOverflowCounter++;
+        }
+    } else {
+        // Failed to acquire mutex.
+        if (nrf_rtc_event_pending(RTC_INSTANCE, NRF_RTC_EVENT_OVERFLOW) || (sOverflowCounter & 0x01)) {
+            // Lower priority context is currently incrementing sOverflowCounter variable.
+            overflowCounter = (sOverflowCounter + 2) / 2;
+        } else {
+            // Lower priority context has already incremented sOverflowCounter variable or incrementing is not needed
+            // now.
+            overflowCounter = sOverflowCounter / 2;
+        }
+    }
+
+    return overflowCounter;
+}
+
+uint32_t getRtcCounter() {
+    return nrf_rtc_counter_get(RTC_INSTANCE);
+}
+
+void getOffsetAndCounter(uint32_t* aOffset, uint32_t* aCounter) {
+    uint32_t offset1 = getOverflowCounter();
+
+    __DMB();
 
-// NOTE: hal_timer_init(), hal_timer_millis() and hal_timer_micros() are
-// implemented in OpenThread platform code (alarm.c). We are re-using alarm implementation
-// necessary for OpenThread internal functionality for our internal needs.
-// It already provides a proper 64-bit stable monotonic microsecond counter, so we use that
-// to derive 64-bit milliseconds out of it, as well as use it in rtc HAL.
+    uint32_t rtcValue1 = getRtcCounter();
+
+    __DMB();
+
+    uint32_t offset2 = getOverflowCounter();
+
+    *aOffset  = offset2;
+    *aCounter = (offset1 == offset2) ? rtcValue1 : getRtcCounter();
+}
+
+uint64_t getCurrentTime() {
+    uint32_t offset;
+    uint32_t rtcCounter;
+
+    getOffsetAndCounter(&offset, &rtcCounter);
+
+    return (uint64_t)offset * US_PER_OVERFLOW + ticksToTime(rtcCounter);
+}
+
+} // anonymous
+
+// Particle-specific
+int hal_timer_init(const hal_timer_init_config_t* conf) {
+    if (conf) {
+        sTimerMicrosBaseOffset = conf->base_clock_offset;
+    }
+
+    sOverflowCounter = 0;
+    sMutex           = 0;
+    sTimeOffset      = 0;
+    sTickCountAtLastOverflow = 0;
+    sTimerMicrosAtLastOverflow = 0;
+
+    // Setup low frequency clock.
+    nrf_drv_clock_lfclk_request(NULL);
+
+    while (!nrf_drv_clock_lfclk_is_running()) {
+        ;
+    }
+
+    // Setup RTC timer.
+    NVIC_SetPriority(RTC_IRQN, RTC_IRQ_PRIORITY);
+    NVIC_ClearPendingIRQ(RTC_IRQN);
+    NVIC_EnableIRQ(RTC_IRQN);
+
+    // This is a workaround for a situation we've seen happen:
+    // despite the fact that the RTC should be stopped and prescaler should be R/W,
+    // we've seen it stay R/O for some reason :|
+    while (rtc_prescaler_get(RTC_INSTANCE) != 0) {
+        nrf_rtc_prescaler_set(RTC_INSTANCE, 0);
+        __DSB();
+        __ISB();
+    }
+
+    nrf_rtc_event_clear(RTC_INSTANCE, NRF_RTC_EVENT_OVERFLOW);
+    nrf_rtc_event_enable(RTC_INSTANCE, RTC_EVTEN_OVRFLW_Msk);
+    nrf_rtc_int_enable(RTC_INSTANCE, NRF_RTC_INT_OVERFLOW_MASK);
+
+    nrf_rtc_event_clear(RTC_INSTANCE, NRF_RTC_EVENT_COMPARE_0);
+    nrf_rtc_event_disable(RTC_INSTANCE, NRF_RTC_EVENT_COMPARE_0);
```

Comment:
In OT they pass `RTC_EVTEN_COMPARE*_Msk` to `nrf_rtc_event_disable()` – does it matter which constant to use in our configuration? I can see two signatures of `nrf_rtc_event_disable()` in nrf_rtc.h.

Reply:
Good catch.
```
__STATIC_INLINE void nrf_rtc_event_disable(NRF_RTC_Type * p_rtc, uint32_t mask)
{
    p_rtc->EVTENCLR = mask;
}
```

Reply:
Fixed.